includes/Services/GitContentRepoManager.php

```diff
@@ -0,0 +1,194 @@
+ */
+final class GitContentRepoManager {
+    private string $cloneBasePath;
+    private LabkiRepoRegistry $repoRegistry;
+
+    public function __construct(?LabkiRepoRegistry $repoRegistry = null) {
+        global $wgCacheDirectory;
+        $this->cloneBasePath = $wgCacheDirectory
+            ? "{$wgCacheDirectory}/labki-content-repos"
+            : wfTempDir() . '/labki-content-repos';
+
+        $this->repoRegistry = $repoRegistry ?? new LabkiRepoRegistry();
+
+        // Ensure base directory exists
+        if (!is_dir($this->cloneBasePath) && !mkdir($this->cloneBasePath, 0755, true)) {
+            throw new RuntimeException("Failed to create clone directory: {$this->cloneBasePath}");
+        }
+    }
+
+    /**
+     * Ensure a clone of the repository exists and is up to date for a given ref.
+     *
+     * @param string $repoUrl Repository URL (e.g., GitHub HTTPS)
+     * @param string $ref Branch, tag, or commit
+     * @return string The local path to the cloned repository
+     * @throws RuntimeException
+     */
+    public function ensureClone(string $repoUrl, string $ref): string {
+        wfDebugLog('labkipack', "ensureClone() called for URL={$repoUrl} ref={$ref}");
+
+        // This is already done in onMediaWikiServices, but we'll do it again here for safety
+        $gitUrl = UrlResolver::resolveContentRepoUrl($repoUrl);
+        $repoDir = $this->generateRepoDirName($gitUrl, $ref);
+        $localPath = "{$this->cloneBasePath}/{$repoDir}";
+
+        if ($this->isValidGitRepo($localPath)) {
+            wfDebugLog('labkipack', "Repository already exists at {$localPath}, updating...");
+            $this->updateClone($localPath, $ref);
+        } else {
+            wfDebugLog('labkipack', "Cloning new repository {$gitUrl} (ref={$ref}) to {$localPath}");
+            $this->cloneRepo($gitUrl, $localPath, $ref);
+        }
+
+        // Determine current commit
+        $commit = $this->getCurrentCommit($localPath);
+        wfDebugLog('labkipack', "Repository {$repoUrl}@{$safeRef} at commit {$commit}");
+
+        // Register repository in DB (unique per URL+ref)
+        $repoId = $this->repoRegistry->ensureRepoEntry($gitUrl, $ref, [
+            'last_commit' => $commit,
+            'updated_at' => \wfTimestampNow(),
+        ]);
+
+        wfDebugLog('labkipack', "Registered/updated repo {$gitUrl}@{$safeRef} as ID={$repoId->toInt()}");
+        return $localPath;
+    }
+
+    /**
+     * Clone repository at specific ref (branch/tag/commit).
+     */
+    private function cloneRepo(string $gitUrl, string $localPath, string $ref): void {
+        if (is_dir($localPath)) {
+            $this->removeDirectory($localPath);
+        }
+
+        // Use --branch for branch/tag; --depth=1 keeps shallow clone
+        $cmd = sprintf(
+            'git clone --depth 1 --branch %s %s %s 2>&1',
+            escapeshellarg($ref),
+            escapeshellarg($gitUrl),
+            escapeshellarg($localPath)
```

Comment:
idk PHP well enough to know offhand if you can define a varargs function in it, but it might be easier for the `runCommand` function to escape all its args instead of doing it each time we call it